R/resample_data.R

```diff
@@ -28,83 +28,135 @@
 #'
 #'
 #' @export
+#'
+
 resample_data = function(data, N, ID_labels=NULL) {
-  # User didn't provide an N or an ID label, it's clear they just want a regular bootstrap
-  if (missing(N) & is.null(ID_labels)) {
-    N <- nrow(data)
-    return(bootstrap_single_level(data, nrow(data), ID_label=NULL))
-  }
+  # Mask internal outer_level and use_dt arguments from view.
+  .resample_data_internal(data, N, ID_labels)
+}
 
-  # Error handling
-  if (!is.null(ID_labels) & (length(N) != length(ID_labels))) {
-    stop(
-      "If you provide more than one ID_labels to resample data for multilevel data, please provide a vector for N of the same length representing the number to resample at each level."
-    )
-  }
+.resample_data_internal = function(data, N, ID_labels=NULL, outer_level=1, use_dt = NA) {
+  # Handle all the data sanity checks in outer_level so we don't have redundant error
+  # checks further down the recursion.
+  if(outer_level) {
+    # Optional usage of data.table to speed up functionality
+    # Short-circuit on the is.na to only attempt the package load if necessary.
+    if(is.na(use_dt) && requireNamespace("data.table", quietly=T)) {
+      use_dt = 1
+    } else {
+      use_dt = 0
+    }
 
-  if (any(!ID_labels %in% names(data))) {
-    stop(
-      "One or more of the ID labels you provided are not columns in the data frame provided."
-    )
-  }
+    # User didn't provide an N or an ID label, it's clear they just want a regular bootstrap
+    # of N units by row.
+    if (missing(N) & is.null(ID_labels)) {
+      return(bootstrap_single_level(data, dim(data)[1], ID_label=NULL))
+    }
```

Comment:
You can move this to the outer function `resample_data()` above.

Reply:
All the error checking can be on the outside function, and the inner function just does the real work.

Reply:
Then you don't really need the `outer` argument here